### DISCUSSION TITLE
Can I use Mask to create an entire shell script (or executable) from the markdown file?

### DISCUSSION THREAD

**Discussion by LinguineCode** [jacobdeichert/mask#107] [Ideas]
Amazing work on this tool. I just found it today so I have a lot to learn about it. I can think of so many uses for it in my personal and professional life.

I have an idea for a feature that you might like

# The problem
1. It's not easy to convince people to install new things, including `mask`, on their systems
2. It's not easy to convince people to prefix their CLI commands with `mask <mycommand>`

# My idea for a solution

1. `mask generate foo.sh` should output a shell script
2. After that challenge is solved, you can have CI/CD run `mask generate foo.sh` triggered on precommit, or release, or etc.
3. Perhaps the resulting `foo.sh` can even commit to the repo, saving an end user the time of generating it themselves, and they can just cherry-pick the script if they want

Then I wouldn't have to convince anyone to do `brew install mask` or run `mask foo`, they can just run `./foo.sh` natively. All the while, I (the author of the repo) got all the benefits of Mask while developing the script.

**Comment by jacobdeichert** [comment]
Hey! This is similar to mask's very first issue #1 

But instead of generating an all-in-one binary, mask would generate a bash script that knows the command tree structure and how to execute the various script types. So maskfile.md would become an authoring format for bash scripts.

However, if mask went down that path, all of mask's features would likely have to be duplicated in bash. And then we'd probably get followup requests to generate outputs for zsh/fish/etc... script types, so I think this would get a bit unwieldy to maintain. 

If anyone chooses to build a 3rd party `mask-to-bash` tool, I'd happily link to it from our README as I think the idea is really useful, but it's just not something I personally want to maintain.

**Reply by LinguineCode** [reply]
> Hey! This is similar to mask's very first issue #1 
> 
> 
> 
> But instead of generating an all-in-one binary, mask would generate a bash script that knows the command tree structure and how to execute the various script types. So maskfile.md would become an authoring format for bash scripts.
> 
> 
> 
> However, if mask went down that path, all of mask's features would likely have to be duplicated in bash. And then we'd probably get followup requests to generate outputs for zsh/fish/etc... script types, so I think this would get a bit unwieldy to maintain. 
> 
> 
> 
> If anyone chooses to build a 3rd party `mask-to-bash` tool, I'd happily link to it from our README as I think the idea is really useful, but it's just not something I personally want to maintain.

That was insightful thank you. #1 no doubt means you've been thinking about this since the project's inception.

If I, or anyone else, would become interested in creating that 3rd party tools, do you have any tips on how to get started? Do you have to know Rust?

**Reply by jacobdeichert** [reply]
Knowing a bit of rust would probably help! While mask's src is quite small, it does use a 3rd party markdown parsing library to read the maskfile.md and also a CLI library (clap) which handles commands and argument setup, so matching the behaviour of these libraries in bash may be a large chunk of work.
